polaris-migrator/src/migrations/replace-sass-motion/replace-sass-motion.ts

```diff
@@ -1,33 +1,144 @@
-  Node,
-  FunctionNode,
-} from 'postcss-value-parser';
+import postcss, {Plugin} from 'postcss';
+import valueParser from 'postcss-value-parser';
 
 import {POLARIS_MIGRATOR_COMMENT} from '../../constants';
 import {
   NamespaceOptions,
   namespace,
   isSassFunction,
-  hasNumericOperator,
+  isNumericOperator,
+  getFunctionArgs,
+  isTransformableDuration,
 } from '../../utilities/sass';
 import {isKeyOf} from '../../utilities/type-guards';
 
+export default function replaceSassMotion(
+  fileInfo: FileInfo,
+  _: API,
+  options: Options,
+) {
+  return postcss(plugin(options)).process(fileInfo.source, {
+    syntax: require('postcss-scss'),
+  }).css;
+}
+
 const processed = Symbol('processed');
-const DEFAULT_DURATION = 'base';
 
-const durationFuncMap = {
-  none: '--p-duration-0',
-  fast: '--p-duration-100',
-  base: '--p-duration-200',
-  slow: '--p-duration-300',
-  slower: '--p-duration-400',
-  slowest: '--p-duration-500',
+interface PluginOptions extends Options, NamespaceOptions {}
+
+const plugin = (options: PluginOptions = {}): Plugin => {
+  const namespacedDuration = namespace('duration', options);
+
+  return {
+    postcssPlugin: 'replace-sass-motion',
+    Declaration(decl) {
+      // @ts-expect-error - Skip if processed so we don't process it again
+      if (decl[processed]) return;
+
+      const handlers = {
+        transition: handleTransitionProps,
+        'transition-duration': handleTransitionProps,
+        'transition-delay': handleTransitionProps,
+      };
+
+      if (!isKeyOf(handlers, decl.prop)) return;
+
+      /**
+       * A collection of transformable values to migrate (e.g. `decl` lengths, functions, etc.)
+       *
+       * Note: This is evaluated at the end of each visitor execution to determine whether
+       * or not to replace the declaration or insert a comment.
+       */
+      const targets: {replaced: boolean}[] = [];
+      let hasNumericOperator = false;
+
+      const parsedValue = valueParser(decl.value);
+
+      handlers[decl.prop]();
+
+      if (targets.some(({replaced}) => !replaced || hasNumericOperator)) {
+        decl.before(postcss.comment({text: POLARIS_MIGRATOR_COMMENT}));
+        decl.before(
+          postcss.comment({
+            text: `${decl.prop}: ${parsedValue.toString()};`,
+          }),
+        );
+      } else {
+        decl.value = parsedValue.toString();
+      }
+
+      //
+      // Handlers
+      //
+
+      function handleTransitionProps() {
+        parsedValue.walk((node) => {
+          if (node.type === 'word') {
+            if (globalValues.has(node.value)) return;
+
+            if (isNumericOperator(node)) {
+              hasNumericOperator = true;
+              return;
+            }
+
+            const dimension = valueParser.unit(node.value);
+
+            if (!isTransformableDuration(dimension)) return;
+
+            targets.push({replaced: false});
+
+            const duration = `${dimension.number}${dimension.unit}`;
```

Comment:
Without normalising the number here, it's possible to accidentally miss a value. For example, `0.50s` would not map to the key `0.5s` in the map.